runtime/doc/options.txt

```diff
@@ -5205,11 +5205,12 @@ A jump table for the options with a short description can be found at |Q_op|.
 	Note that such processing is done after |:set| did its own round of 
 	unescaping, so to keep yourself sane use |:let-&| like shown above.
 							*shell-powershell*
-	To use powershell: >
+	To use PowerShell: >
 		let &shell = has('win32') ? 'powershell' : 'pwsh'
-		set shellquote= shellpipe=\| shellxquote=
+		set shellquote= shellxquote=
 		set shellcmdflag=-NoLogo\ -NoProfile\ -ExecutionPolicy\ RemoteSigned\ -Command
 		set shellredir=\|\ Out-File\ -Encoding\ UTF8
+		set shellpipe=2>&1\ \|\ Out-File\ -Encoding\ UTF8
```

Comment:
Why is `2>&1` prefixed on `shellpipe` only?

Reply:
I omitted it because I was under impression that, for filtering commands, we should not be including error text together with the filtered data that comes out of standard output. However, reading the manual again, it says 
```
For Win32, the Unix checks are done and additionally "cmd" is checked for, which makes the default ">%s 2>&1
```
So, to be consistent, I will change that to capture error stream as well.

I also noticed that, even with `shellredir` configured, I still can't run stuff like `!Select-String` because Vim turns it into `SelectString < filename`. Looks like that pattern is [hardcoded](https://github.com/neovim/neovim/blob/ce976bf3559edca6d3a649c9053fab88959de6cf/src/nvim/ex_cmds.c#L1541), but I don't think it works with PowerShell.

Reply:
I found another potential issue where both `powershell` and `pwsh` seem to have a couple of behavioral quirks:

Both `powershell` and `pwsh` considers a binary to have failed if it writes anything to stderr (it appears that `$?=True` is a "success" flag and is equivalent to Unix's `$?=0`)
```
# Writes to stdout; returns 0
PS C:\Users\dtcaciuc\code\neovim\build> .\bin\shell-test.exe REP 3 hello 2>&1; $?
0: hello
1: hello
2: hello
True
# Writes to stderr; returns 0
PS C:\Users\dtcaciuc\code\neovim\build> .\bin\shell-test.exe 2>&1; $?
ready $
False
```
This results in nvim printing `shell returned 1`

Additionall, `powershell` also writes out an exception object alongside with the program output an so running `r !bin/shell-test` results in the following, which is not ideal:
```
bin\shell-test.exe : ready $
At line:1 char:1
+ bin\shell-test.exe 2>&1 | Out-File -Encoding UTF8 C:\Users\dtcaciuc\A ...
+ ~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (ready $ :String) [], RemoteException
    + FullyQualifiedErrorId : NativeCommandError
```

I'm still going through docs to try to see if there's a way to reconcile these.

Reply:
Check `$LASTEXITCODE` for binaries, `$?` for powershell code.

Reply:
Ok, so after digging in this some more, my take on this is that Powershell 5 that comes bundled with Windows has two problematic behaviors
* Redirect commands seem to write data in odd encoding as reported in the issue
* Whenever caller redirects error stream with 2> and if the executed binary writes anything to error stream, it always turns into an error object (even if the content is a warning), which in turn results in an ugly error trace that I'm not able to silence. I can set ` $ErrorActionPreference = 'SilentlyContinue'` but then it completely discards the error  stream output together with the error object.

Unless there's some trick I could not find, I do not think that it's possible to come up with a good config. The error object behaviour be seen in this simple example and I can't figure out how to work around it.

```
PS C:\Users\dtcaciuc\code\neovim\build> powershell -Command '.\bin\shell-test.exe FOO >out.txt 2>&1'
PS C:\Users\dtcaciuc\code\neovim\build> cat .\out.txt
.\bin\shell-test.exe : Unknown first argument: FOO
At line:1 char:1
+ .\bin\shell-test.exe FOO >out.txt 2>&1
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (Unknown first argument: FOO:String) [], RemoteException
    + FullyQualifiedErrorId : NativeCommandError

```

On the other hand, PowerShell Core (in my case, 7.1.3) seems to behave as expected by default. It produces encoding that Vim understands as-is and it also doesn't seem to automatically turn any error stream content into an error object. Now, the confusing part about this is from what I can gather, it still SHOULD. There's an [experimental feature](https://docs.microsoft.com/en-us/powershell/scripting/learn/experimental-features?view=powershell-7.1#psnotapplyerroractiontostderr) that supposedly needs to be enabled explicitly , but in my case it's off, so I'm not sure what is happening here. Here's an illustration 

```
PS C:\Users\dtcaciuc\code\neovim\build> pwsh -Command 'Get-ExperimentalFeature PSNotApplyErrorActionToStderr'

Name                                Enabled Source                              Description
----                                ------- ------                              -----------
PSNotApplyErrorActionToStderr         False PSEngine                            Don't have $ErrorActionPreference affect stderr …

PS C:\Users\dtcaciuc\code\neovim\build> pwsh -Command '.\bin\shell-test.exe FOO >out.txt 2>&1'
PS C:\Users\dtcaciuc\code\neovim\build> cat .\out.txt
Unknown first argument: FOO
```


Overall, the following is the configuration that works for both PS and PS Core. For PS, this does not tee make output to command window and it produces ugly error traces, but it's functional. For PS Core, it improves on PS behaviour by not printing error objects to output.
 
```
" PowerShell
let &shell = 'powershell'
set shellquote= shellxquote=
set shellcmdflag=-NoLogo\ -NoProfile\ -ExecutionPolicy\ RemoteSigned\ -Command
set shellredir=2>&1\ \|\ Out-File\ -Encoding\ UTF8\ %s;\ exit\ $LastExitCode
set shellpipe=2>&1\ \|\ Out-File\ -Encoding\ UTF8\ %s;\ exit\ $LastExitCode
```

Furthermore, here's a configuration that works even better on PS Core. It causes `:make` and `:r  !<executable>` commands behave identically in my Windows and WSL environments.

```
" PowerShell Core
let shell = 'pwsh'
set shellquote= shellxquote=
set shellcmdflag=-NoLogo\ -NoProfile\ -ExecutionPolicy\ RemoteSigned\ -Command
set shellredir=2>&1\ \|\ Out-File\ %s;\ exit\ $LastExitCode
set shellpipe=2>&1\ \|\ Tee-Object\ %s;\ exit\ $LastExitCode
```

I intend to modify the test config and help doc to include the first PowerShell config because I assume that's what's available on base Windows machines and test hosts. Please let me know if I should add the additional PowerShell Core config to help document as well.

Reply:
Added a blurb about PowerShell 7.1+

Looks like this is the change that is causing the difference https://github.com/PowerShell/PowerShell/pull/13395

Reply:
I started checking different powershell versions and it looks like v6 doesn't have the error object problem either... Strangely enough, powershell 2 on windows 7 also doesn't seem to have it, only powershell 5 on windows 10 does. Powershell is a pretty frustrating program.

Either way, the 7.1 change above is irrelevant if we use `exit $LastExitCode`, so I don't think there's any point in mentioning it.